### PR TITLE
Home work5

### PR DESCRIPTION
**Pull request by Shtuka1993**
Home work 5

### CODE REVIEW

File: src/web/functions.php [hunk 1]
```diff
@@ -11,7 +11,232 @@
...
+/**
+ * Return airport array filtered by first letter
+ *
+ * @param $array
+ * @param $key
+ * @return array
+ */
+function filterByFirstLetter($array, $key)
+{
+    define('FIRST_LETTER', $key);
```
Comment (yderevianko-cartera) [inline src/web/functions.php]: do not use constant inside function

File: src/web/functions.php [hunk 2]
```diff
@@ -11,7 +11,232 @@
...
+
+/**
+ * Returns first letter of airports array item name
+ *
+ * @param $var
+ * @return mixed
+ */
+function getFirstLetter($var)
+{
+
```
Comment (yderevianko-cartera) [inline src/web/functions.php:45]: not needed empty line before return if it is single line in block

File: src/web/functions.php [hunk 3]
```diff
@@ -11,7 +11,232 @@
...
+ * Callback function for filtering by first letter
+ *
+ * Use constant FIRST_LETTER as key
+ *
+ * @param $var
+ * @return bool
+ */
+function filterFirstLetter($var)
+{
+
```
Comment (yderevianko-cartera) [inline src/web/functions.php]: not needed empty line before return if it is single line in block

File: src/web/functions.php [hunk 4]
```diff
@@ -11,7 +11,232 @@
...
+/**
+ * Sort array by key
+ *
+ * @param $array
+ * @param $key
+ * @return mixed
+ */
+function sortByKey($array, $key)
+{
+    usort($array, 'sortBy'.ucfirst($key));
```
Comment (yderevianko-cartera) [inline src/web/functions.php]: should be spaces before and after '.'

File: src/web/functions.php [hunk 5]
```diff
@@ -11,7 +11,232 @@
...
+ */
+function generateURL($request, $key, $value, $resetPage)
+{
+    $request[$key] = $value;
+    if ($resetPage) {
+        $request['page'] = 1;
+    }
+    $result = [];
+    foreach ($request as $key => $value) {
+        $result[] = $key.'='.$value;
```
Comment (yderevianko-cartera) [inline src/web/functions.php]: should have spaces before and after '.'

File: src/web/functions.php [hunk 6]
```diff
@@ -11,7 +11,232 @@
...
+/**
+ * Return count of page
+ *
+ * @param $array
+ * @param $per_page
+ * @return float|int
+ */
+function pagination($array, $per_page)
+{
+    $pages = ( count($array) / $per_page );
```
Comment (yderevianko-cartera) [inline src/web/functions.php]: not needed spaces before ) and after (

File: src/web/index.php [hunk 7]
```diff
@@ -72,10 +103,18 @@
     <table class="table">
         <thead>
         <tr>
-            <th scope="col"><a href="#">Name</a></th>
-            <th scope="col"><a href="#">Code</a></th>
-            <th scope="col"><a href="#">State</a></th>
-            <th scope="col"><a href="#">City</a></th>
+            <th scope="col"><a
```
Comment (yderevianko-cartera) [inline src/web/index.php]: move <a to same line as href

File: src/web/index.php [hunk 8]
```diff
@@ -92,14 +131,16 @@
...
-            <td><a href="#"><?= $airport['state'] ?></a></td>
-            <td><?= $airport['city'] ?></td>
-            <td><?= $airport['address'] ?></td>
-            <td><?= $airport['timezone'] ?></td>
-        </tr>
+            <tr>
+                <td><?= $airport['name'] ?></td>
+                <td><?= $airport['code'] ?></td>
+                <td>
+                    <a href="<?= generateUrl($request, 'filter_by_state', $airport['state'], true) ?>"><?= $airport['state'] ?></a>
```
Comment (yderevianko-cartera) [inline src/web/index.php]: please use <?php echo instead of <?=

File: src/web/index.php [hunk 9]
```diff
@@ -115,9 +156,37 @@
...
+                                                                                  href="<?= generateURL($request, 'page', $i, false) ?>"><?= $i ?></a>
+                    </li>
+                <?php } ?>
+                <?php if ($page <= ($pagesCount - PER_PAGE - 2)) { ?>
+                    <li class="page-item">...</li>
+                <?php } ?>
+                <li class="page-item<?= ($page == $pagesCount) ? ' active' : '' ?>"><a class="page-link"
+                                                                                       href="<?= generateURL($request, 'page', $pagesCount, false) ?>"><?= $pagesCount ?></a>
+                </li>
+            <? } else if ($pagesCount > 1) {
```
Comment (yderevianko-cartera) [inline src/web/index.php]: elseif instead of else if

File: src/web/index.php [hunk 10]
```diff
@@ -115,9 +156,37 @@
...
+            <?php if ($pagesCount > PER_PAGE) { ?>
+                <?php if ($page > (PER_PAGE + 2)) { ?>
+                    <li class="page-item">...</li>
+                <?php } ?>
+                <?php
+                $start = (($page - PER_PAGE) <= 1) ? 2 : ($page - PER_PAGE);
+                $end = (($page + PER_PAGE) >= $pagesCount) ? $pagesCount - 1 : ($page + PER_PAGE);
+                for ($i = $start; $i <= $end; $i++) {
+                    ?>
+                    <li class="page-item<?= ($i == $page) ? ' active' : '' ?>"><a class="page-link"
```
Comment (yderevianko-cartera) [inline src/web/index.php]: move <a to same line as href

File: tests/GetUniqueFirstLettersTest.php [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+     * @dataProvider positiveDataProvider
+     */
+    public function testPositive($input, $expected)
+    {
+        $this->assertEquals($expected, getUniqueFirstLetters($input));
+    }
+
+    public function positiveDataProvider()
+    {
+        return [
```
Comment (yderevianko-cartera) [inline tests/GetUniqueFirstLettersTest.php:17]: Could you please add more positive test cases and if possible negative test cases.
Also it is good idea to test all other functions.

**Review by yderevianko-cartera — CHANGES_REQUESTED** [review]

File: src/web/index.php (hunk 7 above)
Comment (Shtuka1993) [reply src/web/index.php]: Updated

File: src/web/index.php (hunk 8 above)
Comment (Shtuka1993) [reply src/web/index.php]: OK.  I have updated code. I have a question, why we shouldn't use short tags? Is it related to different setting of config files?

Updated.

File: src/web/index.php (hunk 10 above)
Comment (Shtuka1993) [reply src/web/index.php]: Updated

File: src/web/index.php (hunk 9 above)
Comment (Shtuka1993) [reply src/web/index.php]: Updated

File: src/web/functions.php (hunk 1 above)
Comment (Shtuka1993) [reply src/web/functions.php]: Replaced constant with global variable.

File: src/web/functions.php (hunk 2 above)
Comment (Shtuka1993) [reply src/web/functions.php:45]: Updated

File: src/web/functions.php (hunk 3 above)
Comment (Shtuka1993) [reply src/web/functions.php]: Updated

File: src/web/functions.php (hunk 4 above)
Comment (Shtuka1993) [reply src/web/functions.php]: Updated

File: src/web/functions.php (hunk 5 above)
Comment (Shtuka1993) [reply src/web/functions.php]: Updated

File: src/web/functions.php (hunk 6 above)
Comment (Shtuka1993) [reply src/web/functions.php]: Updated

File: src/web/functions.php (hunk 1 above)
Comment (Shtuka1993) [reply src/web/functions.php]: Removed global variable. Used Supper-global $_GET.

File: tests/GetUniqueFirstLettersTest.php (hunk 11 above)
Comment (Shtuka1993) [reply tests/GetUniqueFirstLettersTest.php:17]: Added additional positive cases.  Implemented negative cases test. Updated code of function to handle exceptions.

File: src/web/functions.php [hunk 12]
```diff
@@ -11,7 +11,240 @@
  */
 function getUniqueFirstLetters(array $airports)
 {
-    // put your logic here
+    if (is_array($airports)) {
+        var_dump($airports);//...
```
Comment (yderevianko-cartera) [inline src/web/functions.php]: please remove all var_dump

**Review by yderevianko-cartera — APPROVED** [review]

File: src/web/functions.php (hunk 12 above)
Comment (Shtuka1993) [reply src/web/functions.php]: Removed)
